data_processing/preprocess_text.py

```diff
@@ -0,0 +1,141 @@
+##########
+# Preprocesses Openstax and Life Biology textbooks using Spacy and saves to disk. 
+# Creates separate sentence and key term files for each.
+# For Life, does additional parsing to extract out the knowledge base lexicon and first 10 chapters.
+##########
+import stanfordnlp
+from spacy_stanfordnlp import StanfordNLPLanguage
+from data_processing_utils import write_spacy_docs
+from data_processing_constants import OPENSTAX_TEXTBOOKS, EXCLUDE_SECTIONS
+import warnings
+from io import StringIO
+import pandas as pd
+import re
+import json
+
+def parse_openstax_terms(key_term_text):
+    """ Parse openstax data to extract key terms (including acronyms). """
+    if ":" not in key_term_text:
+        return []
+    term = key_term_text.split(":")[0]
+    match = re.match(".*\((.+)\).*", term)
+    if match:
+        acronym = match.group(1)
+        term = term.replace(f"(acronym)", "")
+        return [term.strip(), acronym.strip()]
+    
+    return [term.strip()] 
+
+def process_lexicon(lexicon):
+    """ Takes in a lexicon consisting of concept text representation pairs and turns this into a 
+    list of Spacy processed terms and a lexicon csv mapping KB concepts to lists of text 
+    representations and their lemma forms.
+    """
+    
+    # get rid of extra column and read in as dataframe
+    lexicon = pd.read_csv(StringIO(lexicon), sep="\s*\|\s*", header=None, 
+                          names=["concept", "relation", "text", "pos"])
+
+    # create mapping from kb concept to unique text representations
+    lexicon = lexicon[~lexicon.text.str.contains("Concept-Word-Frame")]
+    lexicon = lexicon.groupby("concept")["text"].apply(lambda x: list(set(x))).reset_index()
+
+    # spacy process terms to get lemmas
+    spacy_terms = []
+    lemmas = []
+    for concept in lexicon.concept:
+        terms = list(lexicon.loc[lexicon.concept == concept, "text"])[0]
+        terms = [t.replace('"', "").strip() for t in terms]
+        spacy_terms_tmp = [nlp(term) for term in terms]
```

Comment:
do we want to lower case as well for the text?

Reply:
lemmatization will lower case the text unless it thinks it is a proper noun so capitals are handled when matching text. For the lexicon that Vinay provides, everything is already lower case unless it is a proper noun, in which case it is nice to have it capitalized.